src/Storage/Notmuch.hs

```diff
@@ -1,39 +1,86 @@
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE LambdaCase #-}
 -- | module for integrating notmuch within purebred
 module Storage.Notmuch where
 
 import Types (NotmuchMail(..))
 import Notmuch
 import Notmuch.Search
 
-import           Data.Maybe     (fromMaybe)
-import qualified Data.Vector    as Vec
-import           System.Process (readProcess)
+import Control.Monad (void)
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Except (runExceptT)
+import Data.Traversable (traverse)
+import Data.List (union)
+import Data.Maybe (fromMaybe)
+import qualified Data.Vector as Vec
+import System.Process (readProcess)
+import Control.Exception (bracket)
 import qualified Data.Text as T
-import Data.Text.Encoding (decodeUtf8)
-import Types (NotmuchSettings, nmDatabase, mailTags)
+import Data.Text.Encoding (decodeUtf8, encodeUtf8)
+import Types (NotmuchSettings, nmDatabase, mailId, mailTags)
 import Control.Lens.Getter (view)
+import Control.Lens.Setter (set)
 
 
 -- | creates a vector of parsed mails from a not much search
 -- Note, that at this point in time only free form searches are supported. Also,
 -- we filter out the tag which we use to mark mails as new mails
-getMessages :: T.Text -> NotmuchSettings FilePath -> IO (Vec.Vector NotmuchMail)
-getMessages s settings = do
-  db' <- databaseOpen (view nmDatabase settings)
-  case db' of
-    Left status -> do
-        error $ show status
-    Right db -> do
-        q <- query db (FreeForm $ T.unpack s)
-        msgs <- messages q
-        mails <- mapM messageToMail msgs
-        return $ Vec.fromList mails
+getMessages :: T.Text
+            -> NotmuchSettings FilePath
+            -> IO (Either String (Vec.Vector NotmuchMail))
+getMessages s settings = runExceptT
+  (do msgs <- databaseOpenReadOnly (view nmDatabase settings)
+              >>= flip query (FreeForm $ T.unpack s)
+              >>= messages
+      mails <- liftIO $ mapM messageToMail msgs
+      return $ Vec.fromList mails
+  ) >>= pure . either (Left . show) pure
+
+setNotmuchMailTags
+    :: FilePath
+    -> NotmuchMail
+    -> IO (Either String NotmuchMail)
+setNotmuchMailTags dbpath m =
+  case mailTagsToNotmuchTags m of
+    Nothing -> pure $ Left "Tags are corrupt"
+    Just nmtags ->
+      bracket (runExceptT (databaseOpen dbpath) >>= either (error . show) pure)
+              (void . runExceptT . databaseDestroy)
+              (\db -> tagsToMessage nmtags m db)
```

Comment:
Yeah, `bracket` is a bit awkward because it forces `IO`; the transformer is unhelpful here :/
Not much we can do about it right now.